cvxpy/tests/test_KKT.py

```diff
@@ -0,0 +1,301 @@
+        sth.check_complementarity(places=4)
+        sth.verify_dual_values(places=4)
+        sth.check_stationary_lagrangian(places=4)
+        return sth
+
+
+class StandardTestSDPs(BaseTest):
+
+    def test_sdp_1min(self):
+        sth = STH.sdp_1('min')
+        sth.solve(solver='SCS')
+        sth.verify_objective(places=2)
+        sth.check_primal_feasibility(places=2)
+        sth.check_complementarity(places=2)
+        sth.check_dual_domains(places=2)
+        sth.check_stationary_lagrangian(places=2)
+        return sth
+
+    def test_sdp_1max(self):
+        sth = STH.sdp_1('max')
+        sth.solve(solver='SCS')
+        sth.verify_objective(places=2)
+        sth.check_primal_feasibility(places=2)
+        sth.check_complementarity(places=2)
+        sth.check_dual_domains(places=2)
+        sth.check_stationary_lagrangian(places=2)
+        return sth
+
+    def test_sdp_2(self):
+        # places is set to 3 rather than 4, because analytic solution isn't known.
+        sth = STH.sdp_2()
+        sth.solve(solver='SCS')
+        sth.verify_objective(places=2)
+        sth.check_primal_feasibility(places=2)
+        sth.verify_primal_values(places=2)
+        sth.check_complementarity(places=2)
+        sth.check_dual_domains(places=2)
+        sth.check_stationary_lagrangian(places=2)
+        return sth
+
+
+class StandardTestPCPs(BaseTest):
+
+    def test_pcp_1(self, places: int = 2):
+        sth = STH.pcp_1()
+        sth.solve(solver='SCS')
+        sth.verify_objective(places)
+        sth.check_primal_feasibility(places)
+        sth.verify_primal_values(places)
+        sth.check_complementarity(places)
+        sth.verify_dual_values(places)
+        sth.check_stationary_lagrangian(places)
+        return sth
+
+    def test_pcp_2(self, places: int = 2):
+        sth = STH.pcp_2()
+        sth.solve(solver='SCS')
+        sth.verify_objective(places)
+        sth.check_primal_feasibility(places)
+        sth.verify_primal_values(places)
+        sth.check_complementarity(places)
+        sth.verify_dual_values(places)
+        sth.check_stationary_lagrangian(places)
+        return sth
+
+    def test_pcp_3(self, places: int = 2):
+        sth = STH.pcp_3()
+        sth.solve(solver='SCS')
+        sth.verify_objective(places)
+        sth.check_primal_feasibility(places)
+        sth.verify_primal_values(places)
+        sth.check_complementarity(places)
+        sth.verify_dual_values(places)
+        sth.check_stationary_lagrangian(places)
+        return sth
+
+
+class StandardTestFlags(BaseTest):
+    """
+    A class with tests for testing out the incorporation of
+    implicit constraints within `check_stationarity_lagrangian`
+    """
+
+    @staticmethod
+    def tf_1() -> STH.SolverTestHelper:
+        """
+        Tests NSD flag
+        """
+        X = cp.Variable(shape=(3,3), NSD=True)
+        obj = cp.Maximize(cp.lambda_min(X))
+        cons = [X == X.T, X[0][1] == 123]
```

Comment:
Don't index into matrix variables by ``X[i][j]``; use ``X[i, j]`` instead. Please fix this here and elsewhere.

Reply:
Oh! So, `NSD/PSD=True` explicitly enforces `X == X.T` as part of the set of constraints?

Reply:
Well, I would say it's _implicit_ rather than _explicit_, but yes, it enforces a constraint ``X == X.T``.

Reply:
Yeah, the tests still pass even after removing `X == X.T` --- but if the constraint is imposed implicitly (i.e. by simply flipping the `symmetric` flag), meaning that there are now two attributes that are active on the variable (`PSD` & `symmetric`), then the validity of the tests is.... a bit off, right? (since only one of the `if-else` is hit, in this case, the `PSD` one, while for correctness we should be checking for both)

Reply:
You should find that if ``X.is_psd()`` then you also have ``X.is_symmetric()``.
```
>>> import cvxpy as cp
>>> X = cp.Variable(shape=(3,3), PSD=True)
>>> X.is_psd()
True
>>> X.is_symmetric()
True
```

Reply:
Yep, that's what I am getting at. Since now `X` has both the attributes `PSD/NSD` and `symmetric` set to True --- we only iterate through the variables once and make the checks assuming that every variable would at most have one attribute set to `True` (which is ~mostly true since CVXPY only allows users to manually set a single parameter).

My question is this: Since we are checking if the gradient corresponding to this variable lies in the dual cone of the imposed implicit constraint --- in this particular example: imposing `symmetric=True` when `PSD/NSD=True` is fair, but is it also necessary that for some vector $v\in$ `PSD` $\implies v\in$ `Skew(M)` --- if not, then we should be checking for BOTH, correct?

Reply:
You're getting at a subtle issue here.

Let $K$ be a convex cone in an inner product space $E$. The dual to $K$ is $K^* = \\{ y : \langle x, y \rangle \geq 0 \forall x \in K\\}$. Note that an inner product appears in this definition

For a pair of convex cones $K_1, K_2$ both contained in an inner product space $E$, if we have $K_1 \subset K_2$ then we have $K_2^* \subset K_1^*$.

It is essential to ask: what inner product space am I using to define the dual cone?
* When $K$ is the PSD cone, we canonically take the dual where $E$ is the space of symmetric matrices.
* When $K$ is the symmetric cone, it makes little sense to use $E$ as the space of symmetric matrices since this would have $K^* = \\{0\\}$. Instead, when $K$ is the symmetric cone, we use $E$ as the space of all square matrices.
* Something to check for yourself: if $K_n$ is the PSD cone for $n \times n$ matrices and you take its dual where $E$ is the space of all $n \times n$ matrices, then you get $K_n^* = \\mathrm{Skew}(n) + K_n$.